src/frontend/hooks/server/maps.ts

```diff
@@ -0,0 +1,134 @@
+import {useMutation, useQuery, useQueryClient} from '@tanstack/react-query';
+import * as FileSystem from 'expo-file-system';
+import * as v from 'valibot';
+
+import {useApi} from '../../contexts/ApiContext';
+import {DOCUMENT_DIRECTORY, selectFile} from '../../lib/file-system';
+
+import {createRefreshTokenStore} from '../refreshTokenStore';
+import noop from '../../lib/noop';
+
+export const MAPS_QUERY_KEY = 'maps';
+
+const CUSTOM_MAPS_DIRECTORY = new URL('maps', DOCUMENT_DIRECTORY).href;
+const DEFAULT_CUSTOM_MAP_FILE_PATH = CUSTOM_MAPS_DIRECTORY + '/default.smp';
+
+const CustomMapInfoSchema = v.object({
+  created: v.pipe(
+    v.string(),
+    v.isoTimestamp(),
+    v.transform(input => new Date(input)),
+  ),
+  name: v.string(),
+  size: v.pipe(v.number(), v.minValue(0)),
+});
+
+export type CustomMapInfo = v.InferOutput<typeof CustomMapInfoSchema>;
+
+const {useRefreshToken, useRefreshTokenActions} = createRefreshTokenStore();
+
+export function useMapStyleJsonUrl() {
+  const api = useApi();
+  const refreshToken = useRefreshToken();
+
+  return useQuery({
+    queryKey: [MAPS_QUERY_KEY, 'stylejson-url', refreshToken],
+    queryFn: async () => {
+      let result = await api.getMapStyleJsonUrl();
+      return result + `?refresh_token=${refreshToken}`;
+    },
+  });
+}
+
+export function useSelectCustomMapFile() {
+  return useMutation({
+    mutationFn: () => {
+      return selectFile({
+        extensionFilters: ['smp'],
+      });
+    },
+  });
+}
+
+export function useImportCustomMapFile() {
+  const queryClient = useQueryClient();
+  const api = useApi();
+  const {refresh} = useRefreshTokenActions();
+
+  return useMutation({
+    mutationFn: async (opts: {uri: string}) => {
+      await FileSystem.moveAsync({
+        from: opts.uri,
+        to: DEFAULT_CUSTOM_MAP_FILE_PATH,
+      });
+
+      const styleUrl = await api.getMapStyleJsonUrl();
+
+      const response = await fetchCustomMapInfo(styleUrl);
+
+      if (!response.ok) {
+        FileSystem.deleteAsync(DEFAULT_CUSTOM_MAP_FILE_PATH, {
+          idempotent: true,
+        }).catch(noop);
+
+        throw new Error('Invalid map file');
+      }
+    },
+    onSuccess: () => {
+      refresh();
+      queryClient.invalidateQueries({
+        queryKey: [MAPS_QUERY_KEY],
+      });
```

Comment:
yeah the tricky part here is that the invalidation we need is controlled by the consuming map view, not react query. for example:

1. We have an initial value of: `http://localhost:8000/maps/style.json`. This is used by the map view, which makes the http request internally to fetch a response. let's say `{ "name": "my-map" }` for the sake of the example.
2. We do something that should change what style.json we get back in 1 (such as add or remove a map). We invalidate the query, but that isn't enough in some cases because you can still get a value of `http://localhost:8000/maps/style.json`, in which case the mapview won't actually get a new response from that URL because of HTTP caching.

In order to actually force the mapview to fully fetch the response of the same url and ignore the cache, we introduce the refresh token in the form of a url query string, which counts as a new url, even though the query string goes unused by the server that we're calling. we did something similar in Mapeo Mobile for other http resources that would have the same issue (e.g. [here](https://github.com/digidem/mapeo-mobile/blob/develop/src/frontend/api.ts))